feat: Adding basic ntp implementation

bradbeam opened the pull request:
Closes https://github.com/autonomy/talos/issues/279

Signed-off-by: Brad Beam <brad.beam@b-rad.info>

andrewrynhard commented on internal/app/init/pkg/system/services/ntpd.go:
```diff
@@ -0,0 +1,52 @@
+
+// ConditionFunc implements the Service interface.
+func (t *NTPd) ConditionFunc(data *userdata.UserData) conditions.ConditionFunc {
+	// what is this?
```
The conditions allow us to do things like delay the start of a service until a file or files exist on disk.

andrewrynhard commented on internal/app/init/pkg/ntpd/ntpd.go:
```diff
@@ -0,0 +1,93 @@
+// https://access.redhat.com/solutions/39194
+// Using the above as reference for setting min/max
+const (
+	MAXPOLL = 1000
```
You will probably get hit by a linter for all caps.

andrewrynhard commented on internal/app/init/pkg/system/services/ntpd.go:
```diff
@@ -0,0 +1,52 @@
+
+// PreFunc implements the Service interface.
+func (t *NTPd) PreFunc(data *userdata.UserData) error {
+	n := ntpd.New(data.Services.NTPd)
```
Just curious, is there a reason we don't just let the `Daemon` do this?

bradbeam replied:
The main reason is I want to ensure on boot ( prior to starting k8s services ) that we have the proper time set and not wait for the first time.tick to trigger ( could be up to ~20min ). 
I'm also staging it for what I'm expecting a proper ntp implementation to do -- set the system time via https://golang.org/pkg/syscall/#Adjtimex. And I think there's a limitation of the starting time being too far out of sync. So doing the inital sync during startup should address this.

andrewrynhard replied:
SGTM

andrewrynhard commented on internal/app/init/pkg/ntpd/ntpd.go:
```diff
@@ -0,0 +1,93 @@
+		select {
+		case <-ticker.C:
+			// Set some variance with how frequently we poll ntp servers
+			if err = n.Query(); err != nil {
```
I think I would prefer `Query` to return a response and then `SetTime` to accept that as a param, but I can live with this. I feel like there is chance for stale data in the struct. Granted the way we are invoking it here prevents that.

andrewrynhard commented on internal/app/init/pkg/system/services/ntpd.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+func (t *NTPd) Start(data *userdata.UserData) error {
+	go t.client.Daemon()
```
Hmm, this is a different way than we are currently running other services. Not that I'm opposed to this but I'm just thinking out loud. The way we did it with `udevd` was to compile `udevd` as a binary and then use the on host "runner": https://github.com/autonomy/talos/blob/master/internal/app/init/pkg/system/services/udevd.go#L40-L60

I'm not opposed to running that as a baked in service (not a container, and not a separate process), but I'm curious if we want to be consistent. Either that or we create the notion of a service that is not a container or another process. Only reason I would want to aim for container or on host process is because when we can see it with `osctl ps` (currently only true for container based service, but we should extend to on host services in the future).

bradbeam commented:
Just pushed up a containerized version. I'm not sure it's complete yet ( need to test and see if there are any system mounts needed ), but wanted to get it out there. 
I'll squash before merge.

bradbeam commented:
Should be good for review now

```
Last login: Sat Mar 23 15:43:00 2019
~
[16:37:55] $ osctl ps
NAMESPACE   ID       IMAGE          PID    STATUS
system      blockd   talos/blockd   1925   RUNNING
system      ntpd     talos/ntpd     1918   RUNNING
system      osd      talos/osd      1908   RUNNING
system      proxyd   talos/proxyd   2427   RUNNING
system      trustd   talos/trustd   1919   RUNNING
~
[16:38:01] $ osctl logs ntpd
2019/03/23 21:26:54.793480 main.go:54: Starting ntpd
2019/03/23 21:26:54.793586 main.go:74: initial query
2019/03/23 21:26:54.820679 main.go:79: &{Time:2019-03-23 21:26:39.392754641 +0000 UTC ClockOffset:-15.414930644s RTT:25.879201ms Precision:59ns Stratum:2 ReferenceID:3494325793 ReferenceTime:2019-03-23 21:20:48.211599743 +0000 UTC RootDelay:122.07µs RootDispersion:8.270263ms RootDistance:21.270898ms Leap:0 MinError:15.401991043s KissCode: Poll:8s}
2019/03/23 21:26:54.820696 main.go:80: Current time2019/03/23 21:26:54.820703 main.go:81: 2019-03-23 21:26:54.820700438 +0000 UTC m=+0.069732367
2019/03/23 21:26:39.392760 main.go:85: Updated time2019/03/23 21:26:39.392767 main.go:86: 2019-03-23 21:26:39.39276559 +0000 UTC m=+0.069752738
2019/03/23 21:35:01.429542 main.go:96: &{Time:2019-03-23 21:35:01.416654991 +0000 UTC ClockOffset:18.601533ms RTT:62.891021ms Precision:953ns Stratum:2 ReferenceID:2164003585 ReferenceTime:2019-03-23 21:12:37.133895247 +0000 UTC RootDelay:38.772583ms RootDispersion:41.137695ms RootDistance:91.969497ms Leap:0 MinError:0s KissCode: Poll:8s}
2019/03/23 21:35:01.429552 main.go:97: Current time2019/03/23 21:35:01.429557 main.go:98: 2019-03-23 21:35:01.429555758 +0000 UTC m=+502.106542896
2019/03/23 21:35:01.416701 main.go:103: Updated time
2019/03/23 21:35:01.416705 main.go:104: 2019-03-23 21:35:01.416703978 +0000 UTC m=+502.106597237
```
